trainer/trainer.py

```diff
@@ -1,346 +1,606 @@
+        return Stage(
+            name=stage_name,
+            datasets=datasets,
+            until_dataset=dataset_name,
+            until_epoch=int(max_epochs) if max_epochs != 'inf' else None
+        )
+
+    def _load_modifiers(self, ymldata:dict) -> List[Modifier]:
+        """Reads
+        ```yml
+        uppercase: 0.05
+        titlecase: 0.05
+        ```
+        """
+        return [
+            Modifier(name, float(ymldata[name]))
+            for name in MODIFIERS.keys()
+            if name in ymldata
+        ]
+
+
+class CurriculumV2Loader(CurriculumV1Loader):
+    """Slightly different curriculum format that can have multiple files per
+    dataset, and puts the modifiers in their own section."""
+
+    def _load_datasets(self, ymldata:dict, basepath:str) -> Dict[str,Dataset]:
+        """Reads
+        ```yml
+        datasets:
+          clean:
+            - a.gz
+            - b.gz
+        ```
+        """
+        return {
+            name: Dataset(name, [
+                os.path.join(basepath, filepath)
+                for filepath in files
+            ])
+            for name, files in ymldata['datasets']
+        }
+
+    def _load_modifiers(self, ymldata:dict) -> List[Modifier]:
+        """Reads
+        ```yml
+        modifiers:
+          - uppercase 0.05
+          - titlecase 0.05
+        ```
+        """
+        return [
+            self._load_modifier(modifier_line)
+            for modifier_line in ymldata.get('modifiers', [])
+        ]
+
+    def _load_modifier(self, line:str) -> Modifier:
+        name, frequency = line.split()
+        return Modifier(name, float(frequency))
+
+
+class CurriculumLoader:
+    """Reads curriculum yaml files. Wrapper that decides which reader to use
+    based on a version number that may be in there.
+    """
+
+    IMPLEMENTATIONS={
+        '1': CurriculumV1Loader,
+        '2': CurriculumV2Loader,
+    }
+
+    def load(self, fh:Union[TextIO,str,dict], *, **kwargs) -> Curriculum:
+        if isinstance(fh, dict):
+            ymldata = fh
+        else:
+            ymldata = yaml.safe_load(fh)
+
+        impl = self.IMPLEMENTATIONS[str(ymldata.get('version', '1'))]()
+        return impl.load(ymldata, **kwargs)
+
+
+class EpochTracker:
+    """Utility to track how many epochs the reader has progressed since the
+    tracker started tracking."""
+    def __init__(self, reader:DatasetReader):
+        self.reader = reader
+        self.epoch_offset = reader.epoch
+        self.line_offset = reader.line
+
+    @property
+    def epoch(self):
+        epoch = self.reader.epoch - self.epoch_offset
+
+        # ... but if the reader is behind on where it was, it hasn't completed
+        # a full epoch yet.
+        if self.reader.line < self.line_offset:
```

Comment:
I am not sure I understand why this? How/why would epoch move forward without further reading?

Reply:
Because at the moment the dataset readers are not reset between stages. If only half of your `dirty` dataset was read during the `start` stage, then in the `mid` stage the rest will be shown first.

The alternative would be to always reset and reshuffle datasets for each stage. That could work if every stage has its own random seed (and one that doesn't overlap with re-iterated datasets, so can't just do `seed+1`) or if we're happy with skipping remainder of datasets at the end of each stage i.e. always start DatasetReader at `prev_stage.epoch + 1` even if it didn't do a full epoch during the last stage.

Reply:
Oh yeah, i understand. In my original implementation i was not taking into account that i can have a smaller first epoch due to it.